A28-xds-traffic-splitting-and-routing.md

```diff
@@ -0,0 +1,1095 @@
+        *   The cache should only contain the responses for the in-process
+            watches. If we keep the item in cache after the watch is cancelled,
+            it will quickly become outdated.
+
+#### Handle responses
+
+When a response is received, the xDS client will
+
+*   Parse and notify all the watchers for this type
+*   Store the updated in cache, to prepare for future watches for the same
+    resource
+
+# Alternatives considered
+
+## Child policy sharing
+
+One child policy instance, including `cds_experimental` and
+`weighted_target_experimental`, can be an action shared by multiple routes.
+
+For example, the same `{"cds": cluster_1}` can be used as direct actions in
+route1 and route2, and as a child policy of `weighted_target` in route3.
+
+Problem with multiple LB policy instances (not sharing)
+
+*   There may be a lot of routes (resulting in a lot of LB instances)
+    *   The routes also carry RPC configurations, which can be different even if
+        the route actions are the same
+    *   It’s quite easy for users to add a route
+    *   One route only contains one path, not a list of paths
+*   The subchannels are not shared across LB policies in Java and Go (so
+    duplicate TCP connections)
+*   The logic in all child policy will be duplicate (potential memory problem)
+
+Problem with sharing
+
+*   It dramatically complicates the implementation
+    *   The child policy will be owned by multiple parents, life cycle and update
+        propagation (e.g. subchannel state change) get complex. And any picker
+        update will need to fork and notify all parents.
+*   When the route changes, the action may be diverged to a different policy (new
+    or existing). Any state (e.g. stickiness) will be lost.
+
+### Solutions
+
+The problem we want to solve immediately is the duplicate TCP connections. We
+can do
+
+*   Share at the top level `xds_routing_experimental` LB policy
+*   Share all cds policy instances
+*   Share subchannels
+
+The first solution is chosen because it’s easy to implement, and it solves
+most of the subchannel duplication problem. Even wIthout subchannel sharing,
+the number of TCP connections will be linear to the number of different
+actions, not the number of routes. Users can create a large amount of routes,
+but the number of different actions to take is usually small.
+
+# Appendix
+
+## JSON service config generated by xDS resolver
+
+For an RDS response with content
+
+```
+{
+  "URL_MAP/1",
+  path: "/service_1/method_1",
+  cluster: {cluster_1}
+},
+{
+  "URL_MAP/2",
+  path: "/service_1/method_2",
+  cluster: {cluster_1}
+},
+{
+  "URL_MAP/3",
+  prefix: "/service_2/method_2",
+  weighted_clusters: [
+    {name: cluster_1, weight: 75},
+    {name: cluster_2, weight: 25}
+  ]
+},
+{
+  "URL_MAP/4",
+  prefix: "/service_2",
+  weighted_clusters: [
+    {name: cluster_1, weight: 75},
+    {name: cluster_2, weight: 25}
+  ]
+},
+{
+  "URL_MAP/5",
+  regex: "^/service_2/method_3$",
```

Comment:
shouldn't it be safe_regex based on above?

Reply:
Updated